connection_manager.go

```diff
@@ -0,0 +1,393 @@
+	if p := cm.pending[key]; p != nil {
+		cm.Unlock()
+		<-p
+		return nil, nil
+	}
+
+	// Mark that we are creating the connection, so that we can release the lock
+	// and readers for other connections pass through while this connection is
+	// being established, without risking that we might create the connection
+	// more than once
+	p := make(chan struct{})
+	cm.pending[key] = p
+	cm.Unlock()
+
+	// Regardless of whether we succeed or fail, release anyone waiting
+	// for us to be done with the attempt to create
+	defer func() {
+		cm.Lock()
+		delete(cm.pending, key)
+		cm.Unlock()
+		close(p)
+	}()
+
+	// Get the current cluster configuration and watch for changes
+	cw, err := cm.provider.WatchCluster(database, cluster)
+	if err != nil {
+		cm.log.Errorf("failed to watch cluster config for %s:%s: %v", database, cluster, err)
+		return nil, err
+	}
+	<-cw.C()
+	cl := cw.Get()
+
+	// create the connection using the cluster configuration
+	conn, err := cm.openConn(cl)
+	if err != nil {
+		cm.log.Errorf("failed to open connection for %s:%s: %v", database, cluster, err)
+		cw.Close()
+		return nil, err
+	}
+
+	// register the connection and monitor the cluter for updates
+	if err := cm.registerConn(cl, cw, conn); err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+// openConn opens a connection to the given cluster
+func (cm *connectionManager) openConn(cl Cluster) (Connection, error) {
+	d := cm.drivers[cl.Type().Name()]
+	if d == nil {
+		cm.log.Errorf("unsupported driver type %s for %s:%s",
+			cl.Type().Name(), cl.Database(), cl.Name())
+		return nil, errTypeUnsupported
+	}
+
+	cfg := proto.Clone(d.ConfigType())
+	if err := cl.Config().Unmarshal(cfg); err != nil {
+		cm.log.Errorf("failed to unmarshal config v%d for %s:%s: %v",
+			cl.Config().Version(), cl.Database(), cl.Name(), err)
+		return nil, err
+	}
+
+	return d.OpenConnection(cfg)
+}
+
+// registerConn registers a newly opened connection to the given database cluster
+func (cm *connectionManager) registerConn(cl Cluster, cw ClusterWatch, conn Connection) error {
+	key := fmtConnKey(cl.Database(), cl.Name())
+
+	cm.Lock()
+	defer cm.Unlock()
+
+	if cm.closed {
+		cw.Close()
+		conn.Close()
+		return errClosed
+	}
+
+	// Register the watch and connection
+	cm.conns[key], cm.watches[key] = conn, cw
+
+	// Spin up a goroutine to monitor the cluster for updates
+	go cm.watchCluster(cl, cw, conn)
+	cm.wg.Add(1)
+
+	return nil
+}
+
+// watchCluster monitors the given cluster for config updates
+func (cm *connectionManager) watchCluster(cl Cluster, cw ClusterWatch, conn Connection) {
+
+	cm.log.Infof("watching %s:%s", cl.Database(), cl.Name())
+	for range cw.C() {
+		cm.reconfigureConn(cw.Get(), conn)
+	}
+
+	cm.cleanupConn(cl.Database(), cl.Name())
+	cm.wg.Done()
+}
+
+// reconfigureConn reconfigures a connection with new cluster configuration
+func (cm *connectionManager) reconfigureConn(cl Cluster, existing Connection) error {
+	key := fmtConnKey(cl.Database(), cl.Name())
```

Comment:
Would it be worth making a few different versions of `fmtConn...` with one top level just taking a `fmtConnKeyFromCluster(cl Cluster)` to save the double call each time?